### PR TITLE
Icon Picker

### PR DESCRIPTION
**Pull request by piitaya**
## Breaking change

No breaking change

## Proposed change

Replace the `icon-input` by `icon-picker` to improve icon selection.
Support for custom iconsets can be added in the future if they provides a `getIconList` function in addition to the `getIcon`.

I made a little demo for this picker. 

https://user-images.githubusercontent.com/5878303/136159991-c07d3a6a-4463-4d66-b6a8-703634c4bb1f.mp4

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```
```
## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for www.home-assistant.io

### CODE REVIEW

File: src/components/ha-icon.ts [hunk 1]
```diff
@@ -17,348 +17,12 @@ import {
   findIconChunk,
   getIcon,
   Icons,
+  mdiDeprecatedIcons,
   MDI_PREFIXES,
   writeCache,
 } from "../data/iconsets";
 import "./ha-svg-icon";
 
-interface DeprecatedIcon {
```
Comment (piitaya) [inline src/components/ha-icon.ts:25]: Moved to `iconsets.ts` because ha-icon-picker use it too.

File: build-scripts/gulp/gen-icons-json.js [hunk 2]
```diff
@@ -116,5 +116,10 @@ gulp.task("gen-icons-json", (done) => {
     JSON.stringify({ version: package.version, parts })
   );
 
+  fs.writeFileSync(
+    path.resolve(OUTPUT_DIR, "iconList.json"),
+    JSON.stringify(meta.map((icon) => icon.name))
```
Comment (bramkragten) [inline build-scripts/gulp/gen-icons-json.js:121]: Let's just put the icons in here that are in the current version, so we also don't have to filter them out later.

File: src/components/ha-icon-picker.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+  private _initedStates = false;
+
+  private _states: string[] = [];
+
+  @property({ type: Boolean }) private _opened = false;
+
+  protected render(): TemplateResult {
+    return html`
+      <vaadin-combo-box-light
```
Comment (bramkragten) [inline src/components/ha-icon-picker.ts:63]: Use `ha-combo-box` instead, so all the input logic is taken care of.
```suggestion
      <ha-combo-box
```

File: src/components/ha-icon-picker.ts [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+      "value-changed",
+      { value },
+      {
+        bubbles: false,
+        composed: false,
+      }
+    );
+  }
+
+  private _filterChanged(ev: CustomEvent): void {
```
Comment (bramkragten) [inline src/components/ha-icon-picker.ts:116]: Do we need custom filter logic? I think the default works fine?

File: src/components/ha-icon-picker.ts [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+      } else {
+        (this.comboBox as any).filteredItems = [filterString];
+      }
+    } else {
+      (this.comboBox as any).filteredItems = [];
+    }
+  }
+
+  public willUpdate() {
+    if (!this._initedStates) {
```
Comment (bramkragten) [inline src/components/ha-icon-picker.ts:134]:
```suggestion
    if (!cachedState) {
```

File: build-scripts/gulp/gen-icons-json.js (hunk 2 above)
Comment (piitaya) [reply build-scripts/gulp/gen-icons-json.js:121]: Is it right to import `.ts` file from src (where the deprecated icon list in located) inside a build-scripts file ?

File: src/components/ha-icon-picker.ts [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+      if (cachedState) {
+        this._states = cachedState;
+        this._initedStates = true;
+      } else {
+        fetch(`/static/mdi/iconList.json`)
```
[the remark this answers is not in the document]
Comment (bramkragten) [reply src/components/ha-icon-picker.ts:139]: No, it should not be in static then, but in `build`

File: build-scripts/gulp/gen-icons-json.js (hunk 2 above)
Comment (bramkragten) [reply build-scripts/gulp/gen-icons-json.js:121]: Was thinking we could sort this list on the `tags` field in `meta.json` So icons with the tag `Home Automation` will be shown first.

```js
icon.tags.includes("Home Automation");
```

**Comment by bramkragten** [comment]
Sorry for the late response,
I'll take another look tomorrow!

**Review by bramkragten — APPROVED** [review]

**Pull request merged**